### DISCUSSION TITLE
V3.0 release

### DISCUSSION THREAD

**Discussion by moranegg** [codemeta/codemeta#306] [Polls]
Hello CodeMeta community,

We would like you to vote between these two options below, as discussed in https://github.com/codemeta/codemeta/pull/258.

1. naming the release 2.2 and labeling all issues and PRs 2.2. Then  we can quickly move to a v3.0 release. T
   -  the drawback here:  Having v2.2 and v3.0 released closely, might create more confusion about which release should be used.
2. changing the PRs to delete the duplicated terms and releasing a v3.0
   -  the drawback here: Not providing backward compatibility with the next release, it's v2.1 or v3.0 (technically it means that you can't use for example a new v3.0 property `hasSourceCode` or adding natively a role to an author while at the same time using an old v2.1 property with the old name (e.g `embargoDate` and `contIntegration`).

Please vote before June 19th 2023.
Feel free to leave a comment with your take on the subject.

If you have a minute, go and introduce yourself in https://github.com/codemeta/codemeta/discussions/304

**Comment by progval** [comment]
I voted for option 2 (no backward compatibility), because replacing a couple of terms seems simple enough not to warrant the extra complexity of three coexisting versions of the codemeta context.

Either way, existing documents will not be affected because they use the v2.0 context URL (`https://doi.org/10.5063/schema/codemeta-2.0`)

**Comment by mbjones** [comment]
I voted for 2.2, but would be supportive of both paths. 

The reason that I voted for 2.2. is that it would introduce the new properties into the 2.x namespaces, and provide an opportunity for us to directly map old deprecated names to their new counterparts with equivalence relations. That way, when the deprecated term is removed in 3.0, there will be a direct machine-readable path for transformation between the 2.2 context and the 3.0 context.  But I could also see just replacing the old term and skipping the deprecation step, going straight to 3.0.

**Reply by progval** [reply]
> into the 2.x namespaces

Codemeta has a single namespace for both v1.0 and v2.0 (`https://codemeta.github.io/terms/`). Are we planning to create a new namespace for v3.0?

> and provide an opportunity for us to directly map old deprecated names to their new counterparts with equivalence relations

What does that mean? Make the v2.2 context define this?

```
"embargoDate": { "@id":"codemeta:embargoEndDate", "@type": "schema:Date" },
"embargoEndDate": { "@id":"codemeta:embargoEndDate", "@type": "schema:Date" },
```

if yes, then readers of documents with a v2.2 context will get `https://codemeta.github.io/terms/embargoEndDate` as a property name while they expect `https://codemeta.github.io/terms/embargoDate`, causing a compatibility breakage.

**Comment by mbjones** [comment]
Oh, I see what you mean. The context identifier for CodeMeta was originally defined as: 

- For 1.0: https://doi.org/10.5063/schema/codemeta-1.0
- For 2.0: https://doi.org/10.5063/schema/codemeta-2.0

That redirects to the actual context file (despite the problems with content negotiation for ld+json that we discussed elsewhere).

We originally intended these as versioned namespaces, and I forgot that we mapped the terms to `https://codemeta.github.io/terms/` for the term URIs. The problem with that is that those term URIs don't resolve to the context file.

You can see this in operation in the [example codemeta.json](https://github.com/codemeta/codemeta/blob/master/codemeta.json) file for the repository itself.

And yes, it is exactly that kind of equivalence mapping I was referring to, but I don't see why it is an error -- it means both terms can be used and will be mapped onto embargoEndDate. I'd have to look more carefully at the JSON-LD expansion to make sure its right, but my intention was that in 2.2 both terms would be usable (one documented as deprecated), and in 3.0 only the preferred term would remain.

**Reply by progval** [reply]
So what you are saying is that writers of Codemeta documents will allowed to write either `embargoDate` or `embargoEndDate` with a v2.2 context; so for them indeed, it will be a smooth transition.

But readers of a Codemeta document should interpret it after expansion, so if they get this document as input:

```json
{
    "@context": "<hypothetical context URL for v2.2>"
    "embargoDate": "2023-07-01"
}
```

then if `<hypothetical context URL for v2.2>` contains this:

```json
{
    "embargoDate": { "@id":"codemeta:embargoEndDate", "@type": "schema:Date" },
    "embargoEndDate": { "@id":"codemeta:embargoEndDate", "@type": "schema:Date" }
}
```

then readers of the document will interpret it like this:

```json
{
    "https://codemeta.github.io/terms/embargoEndDate": "2023-07-01"
}
```

whereas this document:

```json
{
    "@context": "https://doi.org/10.5063/schema/codemeta-2.0"
    "embargoDate": "2023-07-01"
}
```

was interpreted like this:

```json
{
    "https://codemeta.github.io/terms/embargoDate": "2023-07-01"
}
```

So even if writers can use `embargoDate` in both v2.0 and v2.2 contexts, they will have different meanings in each context.

**Comment by dgarijo** [comment]
I voted for the second option, because we are moving towards v3.0 in the end. I hope we can produce a proper changelog specifying the changes between versions (we can even have a separate machine-readable file for the replacements).
Will v3.0 have a new DOI? Or are we going to move towards the namespace I prepared in https://w3id.org/codemeta/ (e.g. https://w3id.org/codemeta/3.0 not yet enabled)

**Comment by mbjones** [comment]
Given the difficulties we encountered with using DOIs with content negotiation, and our prior discussions on this issue, I think moving towards using the namespaces you prepared at w3id would be best and set us up well for having both human-readable and machine-readable endpoints for the context files.

**Reply by moranegg** [reply]
Hi @mbjones, 
Thank you for your input. I think this can be debated more and I would iwsh we had more input on this discussion.
At the moment, the vote was largely for v3.0 release without a deprecated transition.
I can also tell there is not a lot of adoption yet, and going directly to a v3.0 would ease the comprehensibility of using a vocabulary. Many users do not know the behind the scene aspect of it.

If you do not object, we will go forward with this plan.
Cheers,
Morane

**Comment by moranegg** [comment]
Last call for the version vote!
This vote is open to the larger CodeMeta community, feel free to vote if you are a contributor or a user of CodeMeta.

**Comment by moranegg** [comment]
With 80% we will now move towards a V3.0 release without backward compatibility.
Thank you all for your contributions to this discussion.
